nano/node/fair_queue.hpp

```diff
@@ -0,0 +1,290 @@
+			source{ source },
+			channel{ std::move (channel) }
+		{
+		}
+
+		bool alive () const
+		{
+			if (channel)
+			{
+				return channel->alive ();
+			}
+			// Some sources (eg. local RPC) don't have an associated channel, never remove their queue
+			return true;
+		}
+
+		auto operator<=> (origin const &) const = default;
+	};
+
+private:
+	struct entry
+	{
+		using queue_t = std::deque<Request>;
+		queue_t requests;
+
+		size_t priority;
+		size_t max_size;
+
+		entry (size_t max_size, size_t priority) :
+			priority{ priority },
+			max_size{ max_size }
+		{
+		}
+
+		Request pop ()
+		{
+			release_assert (!requests.empty ());
+
+			auto request = std::move (requests.front ());
+			requests.pop_front ();
+			return request;
+		}
+
+		bool push (Request request)
+		{
+			if (requests.size () < max_size)
+			{
+				requests.push_back (std::move (request));
+				return true; // Added
+			}
+			return false; // Dropped
+		}
+
+		bool empty () const
+		{
+			return requests.empty ();
+		}
+
+		size_t size () const
+		{
+			return requests.size ();
+		}
+	};
+
+public:
+	using source_type = origin;
+	using value_type = std::pair<Request, source_type>;
+
+public:
+	size_t size (source_type source) const
+	{
+		auto it = queues.find (source);
+		return it == queues.end () ? 0 : it->second.size ();
+	}
+
+	size_t max_size (source_type source) const
+	{
+		auto it = queues.find (source);
+		return it == queues.end () ? 0 : it->second.max_size;
+	}
+
+	size_t priority (source_type source) const
+	{
+		auto it = queues.find (source);
+		return it == queues.end () ? 0 : it->second.priority;
+	}
+
+	size_t total_size () const
+	{
+		return std::accumulate (queues.begin (), queues.end (), 0, [] (size_t total, auto const & queue) {
+			return total + queue.second.size ();
+		});
+	};
+
+	bool empty () const
+	{
+		return std::all_of (queues.begin (), queues.end (), [] (auto const & queue) {
+			return queue.second.empty ();
+		});
+	}
+
+	size_t queues_size () const
+	{
+		return queues.size ();
+	}
+
+	void clear ()
+	{
+		queues.clear ();
+	}
+
+	/**
+	 * Should be called periodically to clean up stale channels and update queue priorities and max sizes
+	 */
+	bool periodic_update (std::chrono::milliseconds interval = std::chrono::milliseconds{ 1000 * 30 })
+	{
+		if (elapsed (last_update, interval))
+		{
+			last_update = std::chrono::steady_clock::now ();
+
+			cleanup ();
+			update ();
+
+			return true; // Updated
+		}
+		return false; // Not updated
+	}
+
+	/**
+	 * Push a request to the appropriate queue based on the source
+	 * Request will be dropped if the queue is full
+	 * @return true if added, false if dropped
+	 */
+	bool push (Request request, source_type source)
+	{
+		auto it = queues.find (source);
+
+		// Create a new queue if it doesn't exist
+		if (it == queues.end ())
+		{
+			// TODO: Right now this is constant and initialized when the queue is created, but it could be made dynamic
```

Comment:
Doesn't the update function update these dynamically whenever Iit is called?

Reply:
Forgot to remove this comment, it's a leftover from a previous iteration.